In [38]:
# For windows users
import os
os.environ["OMP_NUM_THREADS"] = '1'

In [39]:
# Import required libraries and dependencies
import pandas as pd
import hvplot.pandas
from pathlib import Path
from sklearn.cluster import KMeans
from sklearn.decomposition import PCA
from sklearn.preprocessing import StandardScaler

In [40]:
# Load the data into a Pandas DataFrame
df_market_data = pd.read_csv(
    Path("Resources/crypto_market_data.csv"),
    index_col="coin_id")

# Display sample data
df_market_data.head(10)

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                    
bitcoin                              1.08388                     7.60278   
ethereum                             0.22392                    10.38134   
tether                              -0.21173                     0.04935   
ripple                              -0.37819                    -0.60926   
bitcoin-cash                         2.90585                    17.09717   
binancecoin                          2.10423                    12.85511   
chainlink                           -0.23935                    20.69459   
cardano                              0.00322                    13.99302   
litecoin                            -0.06341                     6.60221   
bitcoin-cash-sv                      0.92530                     3.29641   

                 price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                     
bitcoin                              6.57509                      7.67258   
ethereum                             4.80849                      0.13169   
tether                               0.00640                     -0.04237   
ripple                               2.24984                      0.23455   
bitcoin-cash                        14.75334                     15.74903   
binancecoin                          6.80688                      0.05865   
chainlink                            9.30098                    -11.21747   
cardano                              5.55476                     10.10553   
litecoin                             7.28931                      1.21662   
bitcoin-cash-sv                     -1.86656                      2.88926   

                 price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                      
bitcoin                             -3.25185                      83.51840   
ethereum                           -12.88890                     186.77418   
tether                               0.28037                      -0.00542   
ripple                             -17.55245                      39.53888   
bitcoin-cash                       -13.71793                      21.66042   
binancecoin                         36.33486                     155.61937   
chainlink                          -43.69522                     403.22917   
cardano                            -22.84776                     264.51418   
litecoin                           -17.23960                      27.49919   
bitcoin-cash-sv                    -24.87434                       7.42562   

                 price_change_percentage_1y  
coin_id                                      
bitcoin                            37.51761  
ethereum                          101.96023  
tether                              0.01954  
ripple                            -16.60193  
bitcoin-cash                       14.49384  
binancecoin                        69.69195  
chainlink                         325.13186  
cardano                           156.09756  
litecoin                          -12.66408  
bitcoin-cash-sv                    93.73082

In [41]:
# Generate summary statistics
df_market_data.describe()

price_change_percentage_24h  price_change_percentage_7d  \
count                    41.000000                   41.000000   
mean                     -0.269686                    4.497147   
std                       2.694793                    6.375218   
min                     -13.527860                   -6.094560   
25%                      -0.608970                    0.047260   
50%                      -0.063410                    3.296410   
75%                       0.612090                    7.602780   
max                       4.840330                   20.694590   

       price_change_percentage_14d  price_change_percentage_30d  \
count                    41.000000                    41.000000   
mean                      0.185787                     1.545693   
std                       8.376939                    26.344218   
min                     -18.158900                   -34.705480   
25%                      -5.026620                   -10.438470   
50%                       0.109740                    -0.042370   
75%                       5.510740                     4.578130   
max                      24.239190                   140.795700   

       price_change_percentage_60d  price_change_percentage_200d  \
count                    41.000000                     41.000000   
mean                     -0.094119                    236.537432   
std                      47.365803                    435.225304   
min                     -44.822480                     -0.392100   
25%                     -25.907990                     21.660420   
50%                      -7.544550                     83.905200   
75%                       0.657260                    216.177610   
max                     223.064370                   2227.927820   

       price_change_percentage_1y  
count                   41.000000  
mean                   347.667956  
std                   1247.842884  
min                    -17.567530  
25%                      0.406170  
50%                     69.691950  
75%                    168.372510  
max                   7852.089700

In [42]:
# Plot your data to see what's in your DataFrame
df_market_data.hvplot.line(
    width=800,
    height=400,
    rot=90
)

:NdOverlay   [Variable]
   :Curve   [coin_id]   (value)

---

### Prepare the Data

In [43]:
# Use the `StandardScaler()` module from scikit-learn to normalize the data from the CSV file
# df_stocks_scaled = StandardScaler().fit_transform(df_stocks[["MeanOpen", "MeanHigh", "MeanLow", "MeanClose", "MeanVolume", "AnnualReturn", "AnnualVariance"]])

df_market_data_scaled = StandardScaler().fit_transform(df_market_data[["price_change_percentage_24h", 
                                                                       "price_change_percentage_7d", 
                                                                       "price_change_percentage_14d",
                                                                       "price_change_percentage_30d",
                                                                       "price_change_percentage_60d",
                                                                       "price_change_percentage_200d",
                                                                       "price_change_percentage_1y"
                                                                      ]])

In [44]:
# Create a DataFrame with the scaled data

df_market_data_scaled = pd.DataFrame(
    df_market_data_scaled,
    columns=["price_change_percentage_24h", 
             "price_change_percentage_7d", 
             "price_change_percentage_14d",
             "price_change_percentage_30d",
             "price_change_percentage_60d",
             "price_change_percentage_200d",
             "price_change_percentage_1y"]
)

# Copy the index crypto names from the original data
# df_market_data

df_market_data_scaled["coin_id"] = df_market_data.index

# Set the coinid column as index

df_market_data_scaled = df_market_data_scaled.set_index("coin_id")

# Display sample data

df_market_data_scaled.head()

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                 
bitcoin                          0.508529                    0.493193   
ethereum                         0.185446                    0.934445   
tether                           0.021774                   -0.706337   
ripple                          -0.040764                   -0.810928   
bitcoin-cash                     1.193036                    2.000959   

              price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                  
bitcoin                          0.772200                     0.235460   
ethereum                         0.558692                    -0.054341   
tether                          -0.021680                    -0.061030   
ripple                           0.249458                    -0.050388   
bitcoin-cash                     1.760610                     0.545842   

              price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                   
bitcoin                         -0.067495                     -0.355953   
ethereum                        -0.273483                     -0.115759   
tether                           0.008005                     -0.550247   
ripple                          -0.373164                     -0.458259   
bitcoin-cash                    -0.291203                     -0.499848   

              price_change_percentage_1y  
coin_id                                   
bitcoin                        -0.251637  
ethereum                       -0.199352  
tether                         -0.282061  
ripple                         -0.295546  
bitcoin-cash                   -0.270317

---

### Find the Best Value for k Using the Original Data.

In [45]:
# Create a list with the number of k-values from 1 to 11

k = list(range(1,11))

In [46]:
# Create an empty list to store the inertia values

inertia = []

# Create a for loop to compute the inertia with each possible value of k
# Inside the loop:
# 1. Create a KMeans model using the loop counter for the n_clusters
# 2. Fit the model to the data using `df_market_data_scaled`
# 3. Append the model.inertia_ to the inertia list

for i in k:
    k_model = KMeans(n_clusters=i,random_state=0, n_init='auto')
    k_model.fit(df_market_data_scaled)
    inertia.append(k_model.inertia_)

In [47]:
# Create a dictionary with the data to plot the Elbow curve

elbow_data = {"k": k, "inertia": inertia}

# Create a DataFrame with the data to plot the Elbow curve

df_elbow = pd.DataFrame(elbow_data)

# Review DataFrame

df_elbow.head()

k     inertia
0  1  287.000000
1  2  198.571818
2  3  123.190482
3  4   79.022435
4  5   68.243549

In [49]:
# Plot a line chart with all the inertia values computed with 
# the different values of k to visually identify the optimal value for k.

df_elbow_plot = df_elbow.hvplot.line(
    x="k",
    y="inertia",
    title="Elbow Curve",
    xticks=k)

df_elbow_plot

:Curve   [k]   (inertia)

#### Answer the following question: 

**Question:** What is the best value for `k`?

**Answer:** 4

---

### Cluster Cryptocurrencies with K-means Using the Original Data

In [50]:
# Initialize the K-Means model using the best value for k

model = KMeans(n_clusters=4, random_state=1, n_init='auto')

In [51]:
# Fit the K-Means model using the scaled data
# df_market_data_scaled

model.fit(df_market_data_scaled)

KMeans(n_clusters=4, n_init='auto', random_state=1)

In [52]:
# Predict the clusters to group the cryptocurrencies using the scaled data

k_4 = model.predict(df_market_data_scaled)

# Print the resulting array of cluster values.

print(k_4)

[1 1 2 2 1 1 1 1 2 2 0 2 2 1 2 2 2 2 1 2 0 1 2 2 2 2 2 0 1 2 2 2 3 2 0 0 1
 0 2 2 0]


In [53]:
# Create a copy of the DataFrame
# df_market_data_scaled

df_market_data_scaled_predictions = df_market_data_scaled.copy()

In [54]:
# Add a new column to the DataFrame with the predicted clusters

df_market_data_scaled_predictions["market_clusters"] = k_4

# Display sample data
df_market_data_scaled_predictions.head()

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                 
bitcoin                          0.508529                    0.493193   
ethereum                         0.185446                    0.934445   
tether                           0.021774                   -0.706337   
ripple                          -0.040764                   -0.810928   
bitcoin-cash                     1.193036                    2.000959   

              price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                  
bitcoin                          0.772200                     0.235460   
ethereum                         0.558692                    -0.054341   
tether                          -0.021680                    -0.061030   
ripple                           0.249458                    -0.050388   
bitcoin-cash                     1.760610                     0.545842   

              price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                   
bitcoin                         -0.067495                     -0.355953   
ethereum                        -0.273483                     -0.115759   
tether                           0.008005                     -0.550247   
ripple                          -0.373164                     -0.458259   
bitcoin-cash                    -0.291203                     -0.499848   

              price_change_percentage_1y  market_clusters  
coin_id                                                    
bitcoin                        -0.251637                1  
ethereum                       -0.199352                1  
tether                         -0.282061                2  
ripple                         -0.295546                2  
bitcoin-cash                   -0.270317                1

In [82]:
# Create a scatter plot using hvPlot by setting 
# `x="price_change_percentage_24h"` and `y="price_change_percentage_7d"`. 
# Color the graph points with the labels found using K-Means and 
# add the crypto name in the `hover_cols` parameter to identify 
# the cryptocurrency represented by each data point.

df_orig_plot = df_market_data_scaled_predictions.hvplot.scatter(
    x="price_change_percentage_24h",
    y="price_change_percentage_7d",
    by="market_clusters",
    hover_cols="coin_id",
    title="Cluster Cryptocurrencies with K-means Using the Original Data")

df_orig_plot

:NdOverlay   [market_clusters]
   :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,coin_id)

---

### Optimize Clusters with Principal Component Analysis.

In [57]:
# Create a PCA model instance and set `n_components=3`.

pca = PCA(n_components=3)

In [58]:
# Use the PCA model with `fit_transform` to reduce to 
# three principal components.
# df_market_data_scaled

df_market_data_scaled_pca = pca.fit_transform(df_market_data_scaled)

# View the first five rows of the DataFrame. 

df_market_data_scaled_pca[0:5]

array([[-0.60066733,  0.84276006,  0.46159457],
       [-0.45826071,  0.45846566,  0.95287678],
       [-0.43306981, -0.16812638, -0.64175193],
       [-0.47183495, -0.22266008, -0.47905316],
       [-1.15779997,  2.04120919,  1.85971527]])

In [86]:
# Retrieve the explained variance to determine how much information 
# can be attributed to each principal component.

pca.explained_variance_ratio_

array([0.3719856 , 0.34700813, 0.17603793])

In [87]:
import numpy as np
total_explained_variance = np.sum(pca.explained_variance_ratio_)
total_explained_variance

0.8950316570309841

#### Answer the following question: 

**Question:** What is the total explained variance of the three principal components?

**Answer:** 0.8950316570309841

In [60]:
# Create a new DataFrame with the PCA data.

df_market_pca = pd.DataFrame(
    df_market_data_scaled_pca,
    columns=["PCA1","PCA2","PCA3"])

# Copy the crypto names from the original data

df_market_pca["coin_id"] = df_market_data.index

# Set the coinid column as index

df_market_pca = df_market_pca.set_index("coin_id")

# Display sample data

df_market_pca.head()

PCA1      PCA2      PCA3
coin_id                                   
bitcoin      -0.600667  0.842760  0.461595
ethereum     -0.458261  0.458466  0.952877
tether       -0.433070 -0.168126 -0.641752
ripple       -0.471835 -0.222660 -0.479053
bitcoin-cash -1.157800  2.041209  1.859715

---

### Find the Best Value for k Using the PCA Data

In [61]:
# Create a list with the number of k-values from 1 to 11

k_pca = list(range(1,11))

In [62]:
# Create an empy list to store the inertia values

inertia_pca = []

# Create a for loop to compute the inertia with each possible value of k
# Inside the loop:
# 1. Create a KMeans model using the loop counter for the n_clusters
# 2. Fit the model to the data using `df_market_data_scaled_pca`
# 3. Append the model.inertia_ to the inertia list

for i in k_pca:
    k_model_pca = KMeans(n_clusters=i,random_state=2, n_init='auto')
    k_model_pca.fit(df_market_data_scaled_pca)
    inertia_pca.append(k_model_pca.inertia_)

In [63]:
# Create a dictionary with the data to plot the Elbow curve

elbow_data_pca = {"k": k_pca, "inertia": inertia_pca}

# Create a DataFrame with the data to plot the Elbow curve

df_elbow_pca = pd.DataFrame(elbow_data_pca)

# Review the DataFrame

df_elbow_pca.head()

k     inertia
0  1  256.874086
1  2  182.339530
2  3  140.431833
3  4   49.665497
4  5   39.174439

In [95]:
# Plot a line chart with all the inertia values computed with 
# the different values of k to visually identify the optimal value for k.

df_elbow_pca_plot = df_elbow_pca.hvplot.line(
    x="k",
    y="inertia",
    title="PCA Elbow Curve",
    xticks=k)

df_elbow_pca_plot

:Curve   [k]   (inertia)

#### Answer the following questions: 

* **Question:** What is the best value for `k` when using the PCA data?

  * **Answer:** 4


* **Question:** Does it differ from the best k value found using the original data?

  * **Answer:** No

### Cluster Cryptocurrencies with K-means Using the PCA Data

In [96]:
# Initialize the K-Means model using the best value for k

model_pca = KMeans(n_clusters=4, random_state=3, n_init='auto')

In [97]:
# Fit the K-Means model using the PCA data
# df_market_pca

model_pca.fit(df_market_pca)

KMeans(n_clusters=4, n_init='auto', random_state=3)

In [98]:
# Predict the clusters to group the cryptocurrencies using the PCA data

k_4_pca = model_pca.predict(df_market_pca)

# Print the resulting array of cluster values.

print(k_4_pca)

[1 1 0 0 1 1 1 1 1 0 0 0 0 1 0 1 0 0 1 0 0 1 0 0 0 0 0 0 1 0 0 0 3 1 0 0 2
 0 0 0 0]


In [100]:
# Create a copy of the DataFrame with the PCA data

df_market_pca_predictions = df_market_pca.copy()

# Add a new column to the DataFrame with the predicted clusters

df_market_pca_predictions["market_clusters"] = k_4_pca

# Display sample data

df_market_pca_predictions.head()

PCA1      PCA2      PCA3  market_clusters
coin_id                                                    
bitcoin      -0.600667  0.842760  0.461595                1
ethereum     -0.458261  0.458466  0.952877                1
tether       -0.433070 -0.168126 -0.641752                0
ripple       -0.471835 -0.222660 -0.479053                0
bitcoin-cash -1.157800  2.041209  1.859715                1

In [101]:
# Create a scatter plot using hvPlot by setting 
# `x="PCA1"` and `y="PCA2"`. 
# Color the graph points with the labels found using K-Means and 
# add the crypto name in the `hover_cols` parameter to identify 
# the cryptocurrency represented by each data point.

df_pca_plot = df_market_pca_predictions.hvplot.scatter(
    x="PCA1",
    y="PCA2",
    by="market_clusters",
    hover_cols="coin_id",
    title="Cluster Cryptocurrencies with K-means Using the PCA Data")

df_pca_plot

:NdOverlay   [market_clusters]
   :Scatter   [PCA1]   (PCA2,coin_id)

### Visualize and Compare the Results

In this section, you will visually analyze the cluster analysis results by contrasting the outcome with and without using the optimization techniques.

In [102]:
# Composite plot to contrast the Elbow curves
# weekly_mean_magnitude_plot + weekly_count_plot

comp_elbow_plot = (df_elbow_plot + df_elbow_pca_plot).cols(1)
comp_elbow_plot

:Layout
   .Curve.I  :Curve   [k]   (inertia)
   .Curve.II :Curve   [k]   (inertia)

In [103]:
# Composite plot to contrast the clusters

comp_clusters_plot = (df_orig_plot + df_pca_plot).cols(1)
comp_clusters_plot

:Layout
   .NdOverlay.I  :NdOverlay   [market_clusters]
      :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,coin_id)
   .NdOverlay.II :NdOverlay   [market_clusters]
      :Scatter   [PCA1]   (PCA2,coin_id)

#### Answer the following question: 

  * **Question:** After visually analyzing the cluster analysis results, what is the impact of using fewer features to cluster the data using K-Means?

  * **Answer:** After examining the cluster analysis results visually, it appears that reducing the number of features used to cluster the data using K-Means had a significant impact. The clusters are more precise and clustered more tightly. In the original plot, the clustering is more dispersed. However, the PCA data plot shows more tightly grouped clusters.